File: arrays-and-strings/02-check-permutation/csharp/SolutionSorting.cs
// Sorting-based solution.
//
// If two strings are permutations of each other,
// their sorted representations will be identical.
//
// Time complexity: O(n log n)
// Space complexity: O(n)

using System;
using System.Linq;

class SolutionSorting
{
    static bool CheckPermutation(string firstWord, string secondWord)
    {
        if (firstWord.Length != secondWord.Length)
            return false;

        string firstWordSorted = string.Concat(firstWord.ToLower().OrderBy(c => c));
        string secondWordSorted = string.Concat(secondWord.ToLower().OrderBy(c => c));

        return firstWordSorted == secondWordSorted;
    }
    


    static void Main(string[] args)
    {
        Console.WriteLine(CheckPermutation("TomMarvoloRiddle", "IamLordVoldemort"));
        Console.WriteLine(CheckPermutation("1991", "9111"));
        Console.WriteLine(CheckPermutation("Tree", "three"));
        Console.WriteLine(CheckPermutation("AMOR", "roma"));
    }
    
}

File: arrays-and-strings/02-check-permutation/java/SolutionSorting.java
// Sorting-based solution.
//
// If two strings are permutations of each other,
// their sorted representations will be identical.
// Easier to reason about,
// but less efficient than frequency counting approaches.
//
// Time complexity: O(n log n)
// Space complexity: O(n)

import java.util.Arrays;

public class SolutionSorting {
    static boolean checkPermutation(String firstWord, String secondWord){

        char[] firstWordArray = firstWord.toLowerCase().toCharArray();
        char[] secondWordArray = secondWord.toLowerCase().toCharArray();

        Arrays.sort(firstWordArray);
        Arrays.sort(secondWordArray);

        String firstWordSorted = new String(firstWordArray);
        String secondWordSorted = new String(secondWordArray);

         
        return firstWordSorted.equals(secondWordSorted);
    }
    public static void main(String[] args){
     
        System.out.println(checkPermutation("TomMarvoloRiddle", "IamLordVoldemort"));
        System.out.println(checkPermutation("1991", "9111"));
        System.out.println(checkPermutation("Tree", "three"));
        System.out.println(checkPermutation("AMOR", "roma"));
    }

}
